Issue #3082: new filter: SuppressByMessageFilter

Rahulkhinchi03 opened the pull request:
Fixes : #3082

Rahulkhinchi03 commented:
I am getting this error but I checked and found that //violation is mentioned and also I have in Uts.
```
Violation lines for C:\Users\ASUS\Desktop\OpenSource\checkstyle\src\test\resources\com\puppycrawl\tools\checkstyle\filters\suppressbymessagefilter\InputSuppressByMessageFilter.java differ.
missing (2): 24, 25
---
expected   : [23, 24, 25, 29]
but was    : [23, 29]
```

rnveach commented:
> I am getting this error but I checked and found that //violation is mentioned and also I have in Uts.

Did you look at a little closer at the error message? It is not talking about missing `// violation` comments in the input file.

Rahulkhinchi03 commented:
> > I am getting this error but I checked and found that //violation is mentioned and also I have in Uts.
> 
> Did you look at a little closer at the error message? It is not talking about missing `// violation` comments in the input file.

Yes, I tried to remove them from violation to see that are they really violating? but when I removed them it was asking for missing and when I added them as violation, they were asking for removal that they aren't expected as violations.

rnveach commented on src/test/java/com/puppycrawl/tools/checkstyle/filters/SuppressByMessageFilterTest.java:
```diff
@@ -0,0 +1,165 @@
+        checkerConfig.addChild(config);
+
+        verify(checkerConfig, fileName, expectedViolations);
+        verifyFilterWithInlineConfigParser(fileName, expectedViolations);
```
Why are you using 2 verify methods in combination? `WithInlineConfigParser` should be the primary verify over anything else.

Rahulkhinchi03 replied:
Okay. Changes Done.

rnveach commented on src/test/java/com/puppycrawl/tools/checkstyle/filters/SuppressByMessageFilterTest.java:
```diff
@@ -0,0 +1,165 @@
+            "28:9: " + getCheckMessage(RequireThisCheck.class, MSG_VARIABLE, "c", ""),
+        };
+
+        verifySuppressed(config, getPath("InputSuppressByMessageFilter.java"),
```
Why is every test using the same input file when you specify the exact configuration to be used inside the input file?
This makes me believe that each test should be using a different input file otherwise each test will just have the same inputs and the same outputs.

Rahulkhinchi03 replied:
Yes, I have created separate input files for all the Unit Test Cases.

Rahulkhinchi03 replied:
After creating the input files and configuring them with `checkFields` and `checkMethods`, all test cases are passing but now I am getting coverage issues.

I believe that it is happening because I have not configured them in input files? or maybe I have done a mistake in something else.
So I tried to configure `checkFormat` and `messageFormat` in the config file of all input files but then I was getting an error like:
`com.puppycrawl.tools.checkstyle.api.CheckstyleException: cannot initialize module com.puppycrawl.tools.checkstyle.TreeWalker - TreeWalker is not allowed as a parent of com.puppycrawl.tools.checkstyle.filters.SuppressByMessageFilter Please review 'Parent Module' section for this Check in web documentation if Check is standard.`

Rahulkhinchi03 replied:
My input config look something like these for the error:

```
/*
com.puppycrawl.tools.checkstyle.checks.coding.RequireThisCheck
checkFields = true
checkMethods = true
validateOnlyOverlapping = false


SuppressByMessageFilter
checkFormat = null
messageFormat = null


*/
```

shashwatj07 replied:
null must be in brackets ```(null)```

rnveach reviewed (changes requested):
These are most likely causing your issues.

romani commented on src/test/resources/com/puppycrawl/tools/checkstyle/filters/suppressbymessagefilter/InputSuppressByMessageFilter.java:
```diff
@@ -0,0 +1,26 @@
+checkFields = true
+checkMethods = true
+validateOnlyOverlapping = false
+
```
Filter config should be defined in comment also.
We already have such filters that migrated to such model of test input

Rahulkhinchi03 replied:
@romani  sir, https://github.com/checkstyle/checkstyle/pull/11174#discussion_r781769380
I have stated the error I am getting while configuring filter.

romani replied:
Please debug why it happening, unfortunately we need to fix that depending problem

romani reviewed (changes requested):
Items:

shashwatj07 commented on src/test/java/com/puppycrawl/tools/checkstyle/filters/SuppressByMessageFilterTest.java:
```diff
@@ -0,0 +1,164 @@
+        final DefaultConfiguration checkerConfig = createRootConfig(treewalkerConfig);
+        checkerConfig.addChild(config);
+
+        verify(checkerConfig, fileName, expectedViolations);
```
This is a forbidden method. Please use parser.

Rahulkhinchi03 replied:
I used parser and now I am getting this error:
`com.puppycrawl.tools.checkstyle.api.CheckstyleException: cannot initialize module com.puppycrawl.tools.checkstyle.TreeWalker - TreeWalker is not allowed as a parent of com.puppycrawl.tools.checkstyle.filters.SuppressByMessageFilter Please review 'Parent Module' section for this Check in web documentation if Check is standard.`

Rahulkhinchi03 replied:
I took a reference from https://github.com/checkstyle/checkstyle/pull/10955
Should I have to use verifyFilterWithInlineConfigParser?

shashwatj07 replied:
> Should I have to use verifyFilterWithInlineConfigParser?

yes

Rahulkhinchi03 replied:
Can you tell me what is suppressed violation?

Rahulkhinchi03 replied:
Do I need to remove the `verifySuppressed` method and add `verifyFilterWithInlineConfigParser` in the UTs or I have to 
just use the parser in the `verifySuppressed` Method?

shashwatj07 replied:
You can do either of them.

> just use the parser in the verifySuppressed Method?

I'll suggest this to avoid repeatition.

Rahulkhinchi03 replied:
But in that case, I have to make `unfiltered violations` or how can I use my `checkerConfig` in the parser?

Rahulkhinchi03 replied:
when I make it like:
`verifyFilterWithInlineConfigParser(fileName, expectedViolations);`
I find these issue: https://github.com/checkstyle/checkstyle/pull/11174#discussion_r787415560

Rahulkhinchi03 replied:
@romani please help!

Rahulkhinchi03 replied:
@rnveach please help!

Rahulkhinchi03 replied:
@nmancus1  need your help.

Rahulkhinchi03 replied:
@romani  sir, please have a look at this and let me know how can I resolve this? https://github.com/checkstyle/checkstyle/pull/11174#discussion_r788918331

rnveach replied:
Have you tried looking at how other tests used inline config parser?
2e46a394a0c2f86b15d3951fed17c50c814b38c3

Rahulkhinchi03 replied:
@rnveach @romani I have seen and understood that I need to add the `removeSuppressed` method but can you please tell me more about the removeSuppressed Method so that I can add this to my code.

shashwatj07 commented:
@Rahulkhinchi03 mention me if you are stuck with parser issues

romani commented:
@Rahulkhinchi03 , please find time to do update

Rahulkhinchi03 commented:
> @Rahulkhinchi03 , please find time to do update

Kindly give your suggestions here: https://github.com/checkstyle/checkstyle/pull/11174#discussion_r782309108

Rahulkhinchi03 commented:
@shashwatj07 please help!

shashwatj07 commented:
Please remove forbidden method invocations and use parser

romani commented:
@Rahulkhinchi03, please reply done to review item that you addressed, please fix CI failures.
Or you wait for us to help with something ?

Rahulkhinchi03 commented:
@romani  sir, I waited for the help but I believe that the mentors are busy with other issues and then @Vyom-Yadav  took this issue to resolve.

Vyom-Yadav commented:
@Rahulkhinchi03 My MSTs are going on, if your doubts get resolved feel free to continue this PR, I haven't worked on this much at present.

rnveach commented:
@Rahulkhinchi03 Please let us know if help is still needed and where.

Rahulkhinchi03 commented:
> @Rahulkhinchi03 Please let us know if help is still needed and where.

when I make it like:
`verifyFilterWithInlineConfigParser(fileName, expectedViolations);`
I find these issue: https://github.com/checkstyle/checkstyle/pull/11174#discussion_r787415560

romani commented:
Please use old style testing, while we have problems with inlined config for filters
